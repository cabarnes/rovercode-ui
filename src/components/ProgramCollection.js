import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import {
  Box,
  Container,
  Grid,
  Button,
  Typography,
  Chip,
  TextField,
  InputAdornment,
  Menu,
  MenuItem,
  Card,
  CardContent,
  CardHeader,
  CardActionArea,
  CardActions,
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import {
  Add,
  ArrowDownward,
  ArrowUpward,
  Search,
} from '@material-ui/icons';
import { Pagination, Autocomplete } from '@material-ui/lab';
import { Link } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';

class ProgramCollection extends Component {
  constructor(props) {
    super(props);

    this.state = {
      page: 1,
      searchQuery: null,
      ordering: 'name',
      tagFilters: [],
      sortMenuAnchorElement: null,
    };
  }

  update = () => {
    const { onUpdate } = this.props;
    const {
      page,
      ordering,
      searchQuery,
      tagFilters,
    } = this.state;

    const params = {
      page,
      ordering,
      tag: tagFilters.join(),
    };

    if (searchQuery) {
      params.search = searchQuery;
    }

    onUpdate(params);
  }

  toggleOrdering = (name) => {
    const { ordering } = this.state;

    if (ordering.endsWith(name)) {
      if (ordering.startsWith('-')) {
        return ordering.substr(1);
      }

      return `-${ordering}`;
    }

    return name;
  }

  handleSortClick = (event) => this.setState({
    sortMenuAnchorElement: event.currentTarget,
  })

  handleSortClose = () => this.setState({
    sortMenuAnchorElement: null,
  })

  handlePageChange = (e, page) => this.setState({
    page,
  }, () => this.update())

  handleSearchChange = (event) => this.setState({
    searchQuery: event.target.value,
    page: 1,
  }, () => this.update())

  handleNameOrderingChange = () => this.setState({
    ordering: this.toggleOrdering('name'),
  }, () => this.update())

  handleTagFilterChange = (event, value) => this.setState({
    tagFilters: value,
  }, () => this.update())

  render() {
    const {
      user,
      label,
      onProgramClick,
      onRemoveClick,
      programs,
      intl,
      tag,
    } = this.props;
    const { ordering, tagFilters, sortMenuAnchorElement } = this.state;

    const searchPlaceholder = intl.formatMessage({
      id: 'app.program_collection.search',
      description: 'Placeholder for search entry',
      defaultMessage: 'Search...',
    });

    const tagFilterPlaceholder = intl.formatMessage({
      id: 'app.program_collection.filter',
      description: 'Placeholder for selecting tags used to filter',
      defaultMessage: 'Tag filters...',
    });

    const sortText = intl.formatMessage({
      id: 'app.program_collection.sort',
      description: 'Button label for sort options',
      defaultMessage: 'Sort',
    });

    const Title = withStyles(() => ({
      root: {
        fontWeight: 'bold',
      },
    }))(Typography);

    const FixedWidthAutocomplete = withStyles(() => ({
      root: {
        width: '100%',
      },
    }))(Autocomplete);

    const TitleArea = withStyles((theme) => ({
      root: {
        marginBottom: theme.spacing(4),
      },
    }))(Grid);

    const PaddedBox = withStyles((theme) => ({
      root: {
        marginTop: theme.spacing(3),
        marginBottom: theme.spacing(3),
      },
    }))(Box);

    return (
      <Container maxWidth="lg">
        <TitleArea item container direction="row" justify="space-between">
          <Grid item>
            <Title variant="h4">
              {label}
            </Title>
          </Grid>
          <Grid item>
            <Button
              variant="contained"
              color="primary"
              size="large"
              startIcon={<Add />}
              component={Link}
              to="/mission-control"
            >
              <FormattedMessage
                id="app.program_list.new"
                description="Button label to create new program"
                defaultMessage="New Program"
              />
            </Button>
          </Grid>
        </TitleArea>
        <Grid container spacing={2}>
          <Grid item>
            <TextField
              variant="outlined"
              size="small"
              className="prompt"
              icon="search"
              placeholder={searchPlaceholder}
              onChange={this.handleSearchChange}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <Search />
                  </InputAdornment>
                ),
              }}
            />
          </Grid>
          <Grid item>
            <Button aria-controls="sort-menu" aria-haspopup="true" onClick={this.handleSortClick}>
              {sortText}
            </Button>
            <Menu
              id="sort-menu"
              anchorEl={sortMenuAnchorElement}
              keepMounted
              open={Boolean(sortMenuAnchorElement)}
              onClose={this.handleSortClose}
              getContentAnchorEl={null}
              anchorOrigin={{
                vertical: 'bottom',
                horizontal: 'center',
              }}
              transformOrigin={{
                vertical: 'top',
                horizontal: 'center',
              }}
            >
              <MenuItem onClick={this.handleNameOrderingChange}>
                <FormattedMessage
                  id="app.program_collection.name"
                  description="Button label to sort by name"
                  defaultMessage="Name"
                />
                {
                  ordering === 'name' ? (
                    <>
                      <ArrowDownward />
                      a ... z
                    </>
                  ) : (
                    <>
                      <ArrowUpward />
                      z ... a
                    </>
                  )
                }
              </MenuItem>
            </Menu>
          </Grid>
          <Grid item xs={2}>
            <FixedWidthAutocomplete
              id="tag-select"
              multiple
              freeSolo
              filterSelectedOptions
              size="small"
              options={tag.tags.map((t) => t.name)}
              onChange={this.handleTagFilterChange}
              value={tagFilters}
              renderInput={(params) => (
                <TextField {...params} variant="outlined" label={tagFilterPlaceholder} />
              )}
              renderTags={(value, getTagProps) => value.map((option, index) => (
                <Chip color="secondary" size="small" label={option} {...getTagProps({ index })} />
              ))}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3}>
          {
            programs.results.map((program) => (
              <Grid item xs={12} md={6} lg={3}>
                <Card key={program.id}>
                  <CardHeader
                    title={program.name}
                    subheader={
                        user.username === program.user.username ? (
                          <FormattedMessage
                            id="app.program_collection.mine"
                            description="Label to indicate program owned by user"
                            defaultMessage="Mine"
                          />
                        ) : program.user.username
                      }
                  />
                  {/* <CardActionArea>
                    <CardContent>
                      Lorem ipsum dolor sit amet, consectetur adipiscing elit,
                      sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                    </CardContent>
                  </CardActionArea> */}
                  <CardActions>
                    <Button
                      id={program.id}
                      data-owned={user.username === program.user.username}
                      onClick={onProgramClick}
                    >
                      {
                        user.username === program.user.username ? (
                          <FormattedMessage
                            id="app.program_collection.work"
                            description="Button label to keep working on program"
                            defaultMessage="Keep Working"
                          />
                        ) : (
                          <FormattedMessage
                            id="app.program_collection.view"
                            description="Button label to view a program"
                            defaultMessage="View"
                          />
                        )
                      }
                    </Button>
                    {
                      user.username === program.user.username ? (
                        <Button
                          id={program.id}
                          name={program.name}
                          onClick={onRemoveClick}
                          floated="right"
                        >
                          <FormattedMessage
                            id="app.program_collection.remove"
                            description="Button label to remove a program"
                            defaultMessage="Remove"
                          />
                        </Button>
                      ) : (null)
                    }
                  </CardActions>
                </Card>
              </Grid>
            ))
          }
        </Grid>
        {
          programs.total_pages > 1 ? (
            <PaddedBox
              display="flex"
              justifyContent="center"
            >
              <Pagination
                defaultPage={1}
                count={programs.total_pages}
                showFirstButton
                showLastButton
                color="secondary"
                onChange={this.handlePageChange}
              />
            </PaddedBox>
          ) : (null)
        }
      </Container>
    );
  }
}

ProgramCollection.defaultProps = {
  tag: {
    tags: [],
  },
};

ProgramCollection.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }).isRequired,
  tag: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  }),
  label: PropTypes.string.isRequired,
  onProgramClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(injectIntl(ProgramCollection));
