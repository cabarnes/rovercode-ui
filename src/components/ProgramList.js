import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import Add from '@material-ui/icons/Add';
import {
  Confirm,
  Icon,
  Loader,
  Segment,
} from 'semantic-ui-react';
import { FormattedMessage, injectIntl } from 'react-intl';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

import ProgramCollection from './ProgramCollection';

const defaultState = {
  programLoaded: false,
  confirmOpen: false,
  focusProgram: {
    id: null,
    name: null,
  },
};

class ProgramList extends Component {
  constructor(props) {
    super(props);

    this.state = defaultState;
  }

  componentDidMount() {
    const {
      clearProgram,
      fetchPrograms,
      fetchTags,
      user,
    } = this.props;

    clearProgram();

    return fetchPrograms({
      user: user.user_id,
    }).then(() => fetchPrograms({
      user__not: user.user_id,
    })).then(() => fetchPrograms({
      admin_tags: 'featured',
    })).then(() => fetchTags());
  }

  showConfirm = (e) => this.setState({
    confirmOpen: true,
    focusProgram: {
      id: e.target.parentNode.id || e.target.id,
      name: e.target.parentNode.name || e.target.name,
    },
  })

  cancelRemove = () => this.setState(defaultState)

  removeProgram = () => {
    const { fetchPrograms, removeProgram, user } = this.props;
    const { focusProgram } = this.state;

    this.setState(defaultState);

    return removeProgram(focusProgram.id)
      .then(() => fetchPrograms({
        user: user.user_id,
      }))
      .then(() => fetchPrograms({
        user__not: user.user_id,
      }))
      .then(() => fetchPrograms({
        admin_tags: 'featured',
      }));
  }

  loadProgram = (e) => {
    const { changeReadOnly, fetchProgram } = this.props;
    let program = e.target;
    if (e.target.parentNode.id) {
      program = e.target.parentNode;
    }
    const readOnly = program.dataset.owned === 'false';


    fetchProgram(program.id).then(() => {
      changeReadOnly(readOnly);
      this.setState({
        programLoaded: true,
      });
    });
  }

  fetchUserPrograms = (params) => {
    const { fetchPrograms, user } = this.props;
    fetchPrograms({
      user: user.user_id,
      ...params,
    });
  }

  fetchFeaturedPrograms = (params) => {
    const { fetchPrograms } = this.props;
    fetchPrograms({
      admin_tags: 'featured',
      ...params,
    });
  }

  fetchOtherPrograms = (params) => {
    const { fetchPrograms, user } = this.props;
    fetchPrograms({
      user__not: user.user_id,
      ...params,
    });
  }

  programSegment = (programs, user, tag, label, onUpdate) => (
    <Segment raised style={{ margin: '10px 10% 10px 10%' }}>
      <ProgramCollection
        programs={programs}
        user={user}
        tag={tag}
        label={label}
        onProgramClick={this.loadProgram}
        onRemoveClick={this.showConfirm}
        onUpdate={onUpdate}
      />
    </Segment>
  )

  render() {
    const {
      intl,
      programs,
      user,
      tag,
      userPrograms,
      featuredPrograms,
    } = this.props;
    const {
      confirmOpen,
      focusProgram,
      programLoaded,
    } = this.state;

    const myProgramsHeader = intl.formatMessage({
      id: 'app.program_list.my_programs',
      description: 'Header for all of user\'s programs',
      defaultMessage: 'My Programs',
    });

    const featuredProgramsHeader = intl.formatMessage({
      id: 'app.program_list.featured_programs',
      description: 'Header for all featured programs',
      defaultMessage: 'Featured Programs',
    });

    const otherProgramsHeader = intl.formatMessage({
      id: 'app.program_list.other_programs',
      description: 'Header for finding other user\'s programs',
      defaultMessage: 'Find More',
    });

    const cancelButtonText = intl.formatMessage({
      id: 'app.program_list.cancel',
      description: 'Button label to cancel removing program',
      defaultMessage: 'No',
    });

    const confirmButtonText = intl.formatMessage({
      id: 'app.program_list.confirm',
      description: 'Button label to confirm removing program',
      defaultMessage: 'Yes',
    });

    const dialogHeader = intl.formatMessage({
      id: 'app.program_list.dialog_header',
      description: 'Header for removing program confirmation dialog',
      defaultMessage: 'Remove Program',
    });

    const dialogContent = intl.formatMessage({
      id: 'app.program_list.dialog_content',
      description: 'Asks the user to confirm removing program',
      defaultMessage: 'Are you sure you want to remove {name}?',
    }, {
      name: focusProgram.name,
    });

    return (
      <>
        <Button variant="contained" color="primary" href="/mission-control" style={{ marginLeft: '10%' }}>
          <Add />
          <FormattedMessage
            id="app.program_list.new"
            description="Button label to create new program"
            defaultMessage="New Program"
          />
        </Button>
        {
          programLoaded ? (
            <Redirect to={{
              pathname: '/mission-control',
            }}
            />
          ) : (null)
        }
        {
          userPrograms === null
            ? (<Loader active />)
            : this.programSegment(userPrograms, user, tag,
              myProgramsHeader, this.fetchUserPrograms)
        }
        {
          featuredPrograms === null
            ? (<Loader active />)
            : this.programSegment(featuredPrograms, user, tag,
              featuredProgramsHeader, this.fetchFeaturedPrograms)
        }
        {
          programs === null
            ? (<Loader active />)
            : this.programSegment(programs, user, tag, otherProgramsHeader, this.fetchOtherPrograms)
        }
        <Confirm
          header={dialogHeader}
          content={dialogContent}
          open={confirmOpen}
          onConfirm={this.removeProgram}
          onCancel={this.cancelRemove}
          cancelButton={cancelButtonText}
          confirmButton={confirmButtonText}
        />
      </>
    );
  }
}

ProgramList.defaultProps = {
  programs: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
  userPrograms: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
  featuredPrograms: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
  tag: {
    tags: [],
  },
};

ProgramList.propTypes = {
  fetchProgram: PropTypes.func.isRequired,
  fetchPrograms: PropTypes.func.isRequired,
  removeProgram: PropTypes.func.isRequired,
  changeReadOnly: PropTypes.func.isRequired,
  fetchTags: PropTypes.func.isRequired,
  clearProgram: PropTypes.func.isRequired,
  user: PropTypes.shape({
    user_id: PropTypes.number.isRequired,
    username: PropTypes.string.isRequired,
  }).isRequired,
  programs: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }),
  userPrograms: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }),
  featuredPrograms: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        user: PropTypes.shape({
          username: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ),
  }),
  tag: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  }),
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default injectIntl(ProgramList);
