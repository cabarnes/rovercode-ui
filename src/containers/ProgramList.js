import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { changeReadOnly as actionChangeReadOnly, clearProgram, fetchProgram } from '../actions/code';
import { fetchPrograms, removeProgram } from '../actions/program';
import { checkAuthError, authHeader } from '../actions/auth';
import { fetchTags } from '../actions/tag';
import ProgramList from '../components/ProgramList';

const mapStateToProps = ({
  code, program, tag, user,
}) => ({
  code, ...program, tag, user,
});
const mapDispatchToProps = (dispatch, { cookies }) => ({
  fetchProgram: (id) => dispatch(fetchProgram(id, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
  fetchPrograms: (params) => {
    const xhrOptions = authHeader(cookies);

    xhrOptions.params = params;

    return dispatch(fetchPrograms(xhrOptions)).catch(checkAuthError(dispatch));
  },
  removeProgram: (id) => dispatch(removeProgram(id, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
  changeReadOnly: (isReadOnly) => dispatch(actionChangeReadOnly(isReadOnly)),
  fetchTags: () => dispatch(fetchTags(authHeader(cookies))).catch(checkAuthError(dispatch)),
  clearProgram: () => dispatch(clearProgram()),
});

const ProgramListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProgramList);

ProgramListContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(ProgramListContainer));
