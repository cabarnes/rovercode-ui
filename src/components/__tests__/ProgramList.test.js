import React from 'react';
import { MemoryRouter } from 'react-router';
import { CircularProgress } from '@material-ui/core';
import { Redirect } from 'react-router-dom';
import ProgramCollection from '../ProgramCollection';
import ProgramList from '../ProgramList';

let changeReadOnly;
let fetchProgram;
let fetchPrograms;
let removeProgram;
let fetchTags;
let clearProgram;
let clearProgramList;

describe('The ProgramList component', () => {
  beforeEach(() => {
    changeReadOnly = jest.fn();
    fetchProgram = jest.fn().mockResolvedValue();
    fetchPrograms = jest.fn().mockResolvedValue();
    removeProgram = jest.fn().mockResolvedValue();
    fetchTags = jest.fn().mockResolvedValue();
    clearProgram = jest.fn();
    clearProgramList = jest.fn();
  });

  test('renders on the page with no errors', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();
    wrapper.setState({
      focusProgram: {
        id: 1,
        name: 'Test Program',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches programs and tags on mount', async () => {
    await mountWithIntl(
      <MemoryRouter>
        <ProgramList
          changeReadOnly={changeReadOnly}
          fetchProgram={fetchProgram}
          fetchPrograms={fetchPrograms}
          removeProgram={removeProgram}
          fetchTags={fetchTags}
          clearProgram={clearProgram}
          clearProgramList={clearProgramList}
          user={{ user_id: 1, username: 'testuser' }}
          owned
        />
      </MemoryRouter>,
    );
    expect(fetchTags.mock.calls.length).toBe(1);
    expect(fetchPrograms.mock.calls.length).toBe(1);
    expect(clearProgram.mock.calls.length).toBe(1);
  });

  test('shows the correct number of programs for the user', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }, {
        id: 5,
        name: 'Unnamed_Design_2',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'testuser',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        userPrograms={programs}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(true);
    expect(wrapper.find(CircularProgress).exists()).toBe(false);
  });

  test('shows loading when programs fetching', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        programs={null}
        owned
      />,
    ).dive().dive();

    expect(wrapper.find(ProgramCollection).exists()).toBe(false);
    expect(wrapper.find(CircularProgress).exists()).toBe(true);
    expect(wrapper.find(CircularProgress).length).toBe(1);
  });

  test('redirects to mission control when program loads', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.setState({
      programLoaded: true,
    });

    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).at(0).prop('to')).toEqual({
      pathname: '/mission-control',
    });
  });

  test('loads a program', (done) => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        programs={programs}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().loadProgram({
      target: {
        parentNode: {
          parentNode: {
            id: undefined,
          },
        },
        id: 33,
        dataset: {
          owned: 'false',
        },
      },
    }).then(() => {
      expect(changeReadOnly).toHaveBeenCalledWith(true);
      expect(fetchProgram).toHaveBeenCalledWith(33);
      expect(wrapper.state('programLoaded')).toBe(true);

      wrapper.setState({
        programLoaded: false,
      });

      wrapper.instance().loadProgram({
        target: {
          parentNode: {
            parentNode: {
              id: 55,
              dataset: {
                owned: 'true',
              },
            },
          },
        },
      }).then(() => {
        expect(changeReadOnly).toHaveBeenCalledWith(false);
        expect(fetchProgram).toHaveBeenCalledWith(33);
        expect(wrapper.state('programLoaded')).toBe(true);
        done();
      });
    });
  });

  test('fetches user programs after page change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().fetchUserPrograms({
      page: 2,
    }, true);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user: 1,
      page: 2,
    });
  });

  test('fetches community programs after page change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned={false}
      />,
    ).dive().dive();

    wrapper.instance().fetchCommunityPrograms({
      page: 2,
    }, false);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user__not: 1,
      page: 2,
    });
  });

  test('fetches programs after search change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().fetchUserPrograms({
      search: 'abc',
      page: 1,
    }, true);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user: 1,
      page: 1,
      search: 'abc',
    });
  });

  test('fetches other programs after search change', () => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned={false}
      />,
    ).dive().dive();

    wrapper.instance().fetchCommunityPrograms({
      search: 'abc',
      page: 1,
    }, false);

    expect(fetchPrograms).toHaveBeenCalledWith({
      user__not: 1,
      page: 1,
      search: 'abc',
    });
  });

  test('removes a program and reloads the program list', (done) => {
    const programs = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 33,
        name: 'Unnamed_Design_3',
        content: '<xml><variables></variables></xml>',
        user: {
          username: 'admin',
        },
      }],
    };
    const wrapper = shallowWithIntl(
      <ProgramList
        programs={programs}
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.setState({
      focusProgram: {
        id: 33,
        name: 'Unnamed_Design_3',
      },
    });
    wrapper.instance().removeProgram().then(() => {
      expect(fetchPrograms).toHaveBeenCalledTimes(2);
      expect(removeProgram).toHaveBeenCalledWith(33);
      done();
    });
  });

  test('shows confirm dialog', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().showConfirm({
      target: {
        parentNode: {
          parentNode: {
            id: undefined,
          },
        },
        id: 33,
        name: 'Unnamed_Design_3',
      },
    });

    expect(wrapper.state('confirmOpen')).toBe(true);
  });

  test('cancel dialog does not remove program', () => {
    const wrapper = shallowWithIntl(
      <ProgramList
        changeReadOnly={changeReadOnly}
        fetchProgram={fetchProgram}
        fetchPrograms={fetchPrograms}
        removeProgram={removeProgram}
        fetchTags={fetchTags}
        clearProgram={clearProgram}
        clearProgramList={clearProgramList}
        user={{ user_id: 1, username: 'testuser' }}
        owned
      />,
    ).dive().dive();

    wrapper.instance().cancelRemove();

    expect(fetchPrograms).toHaveBeenCalledTimes(1);
    expect(removeProgram).not.toHaveBeenCalled();
    expect(wrapper.state('confirmOpen')).toBe(false);
  });
});
