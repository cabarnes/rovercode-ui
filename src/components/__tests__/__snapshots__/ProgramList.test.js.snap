// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramList component renders on the page with no errors 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Button))
    color="primary"
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Link",
        "propTypes": Object {
          "innerRef": [Function],
          "onClick": [Function],
          "replace": [Function],
          "target": [Function],
          "to": [Function],
        },
        "render": [Function],
      }
    }
    size="large"
    startIcon={<Memo(AddIcon) />}
    style={
      Object {
        "marginLeft": "10%",
      }
    }
    to="/mission-control"
    variant="contained"
  >
    <FormattedMessage
      defaultMessage="New Program"
      id="app.program_list.new"
      values={Object {}}
    />
  </WithStyles(ForwardRef(Button))>
  <Segment
    raised={true}
    style={
      Object {
        "margin": "10px 10% 10px 10%",
      }
    }
  >
    <injectIntl(ProgramCollection)
      label="My Programs"
      onProgramClick={[Function]}
      onRemoveClick={[Function]}
      onUpdate={[Function]}
      programs={
        Object {
          "next": null,
          "previous": null,
          "results": Array [],
          "total_pages": 1,
        }
      }
      tag={
        Object {
          "tags": Array [],
        }
      }
      user={
        Object {
          "user_id": 1,
          "username": "testuser",
        }
      }
    />
  </Segment>
  <Segment
    raised={true}
    style={
      Object {
        "margin": "10px 10% 10px 10%",
      }
    }
  >
    <injectIntl(ProgramCollection)
      label="Featured Programs"
      onProgramClick={[Function]}
      onRemoveClick={[Function]}
      onUpdate={[Function]}
      programs={
        Object {
          "next": null,
          "previous": null,
          "results": Array [],
          "total_pages": 1,
        }
      }
      tag={
        Object {
          "tags": Array [],
        }
      }
      user={
        Object {
          "user_id": 1,
          "username": "testuser",
        }
      }
    />
  </Segment>
  <Segment
    raised={true}
    style={
      Object {
        "margin": "10px 10% 10px 10%",
      }
    }
  >
    <injectIntl(ProgramCollection)
      label="Find More"
      onProgramClick={[Function]}
      onRemoveClick={[Function]}
      onUpdate={[Function]}
      programs={
        Object {
          "next": null,
          "previous": null,
          "results": Array [],
          "total_pages": 1,
        }
      }
      tag={
        Object {
          "tags": Array [],
        }
      }
      user={
        Object {
          "user_id": 1,
          "username": "testuser",
        }
      }
    />
  </Segment>
  <Confirm
    cancelButton="No"
    confirmButton="Yes"
    content="Are you sure you want to remove Test Program?"
    header="Remove Program"
    onCancel={[Function]}
    onConfirm={[Function]}
    open={false}
    size="small"
  />
</Fragment>
`;
